orchestration/templates/ingest-file.yaml

```diff
@@ -0,0 +1,153 @@
+apiVersion: argoproj.io/v1alpha1
+kind: WorkflowTemplate
+metadata:
+  name: ingest-file
+spec:
+  templates:
+    - name: main
+      inputs:
+        parameters:
+          - name: api-url
+          - name: dataset-id
+          - name: profile-id
+          - name: gcs-bucket
+          - name: gcs-file-path
+          - name: target-path
+          - name: timeout
+      dag:
+        tasks:
+          # 1) Make sure that something doesn't already exist at the target path
+          - name: check-existing-file
+            template: check-existing-file
+            arguments:
+              parameters:
+                - name: api-url
+                  value: '{{ "{{inputs.parameters.api-url}}" }}'
+                - name: dataset-id
+                  value: '{{ "{{inputs.parameters.dataset-id}}" }}'
+                - name: target-path
+                  value: '{{ "{{inputs.parameters.target-path}}" }}'
+          # 2) Make a post request to the Jade API to ingest the xml blob
+          - name: submit-file-ingest
+            template: submit-file-ingest
+            when: '{{ "tasks.check-existing-file.outputs.result}} == 200" }}'
+            arguments:
+              parameters:
+                - name: api-url
+                  value: '{{ "{{inputs.parameters.api-url}}" }}'
+                - name: dataset-id
+                  value: '{{ "{{inputs.parameters.dataset-id}}" }}'
+                - name: profile-id
+                  value: '{{ "{{inputs.parameters.profile-id}}" }}'
+                - name: source-path
+                  value: gs://{{ "{{inputs.parameters.gcs-bucket}}" }}/{{ "{{inputs.parameters.gcs-file-path}}" }}
+                - name: target-path
+                  value: '{{ "{{inputs.parameters.target-path}}" }}'
+          # 3) Poll the ingest job for the file repeatedly until the file has been ingested
+          - name: poll-ingest-job
+            dependencies: [submit-file-ingest]
+            template: poll-ingest-job
+            arguments:
+              parameters:
+                - name: api-url
+                  value: '{{ "{{inputs.parameters.api-url}}" }}'
+                - name: job-id
+                  value: '{{ "{{tasks.submit-file-ingest.outputs.result}}" }}'
+                - name: timeout
+                  value: '{{ "{{inputs.parameters.timeout}}" }}'
+          # 4) Return the fileId
+          - name: get-job-result
+            dependencies: [poll-ingest-job]
+            template: get-job-result
+            when: '{{ "tasks.poll-ingest-job.outputs.result}} == True" }}'
+            arguments:
+              parameters:
+                - name: api-url
+                  value: '{{ "{{inputs.parameters.api-url}}" }}'
+                - name: job-id
+                  value: '{{ "{{tasks.submit-file-ingest.outputs.result}}" }}'
```

Comment:
This needs two outputs:
1. An output for the result of "check-existing-file"
2. An output for the DRS ID of the new file, if it was ingested

Point 2 will probably fail now. They just merged a change in Argo to allow for setting `default` args on outputs that depend on skipped steps. We'll need to set some dummy value (i.e. empty string), and then have downstream steps only try to use the DRS ID if the check-exists output is false.

Reply:
why do we need the output of "check-existing-file"? Shouldn't the presence or lack thereof of a DRS ID be sufficient?

Reply:
Argo doesn't have optional outputs. If we skip the ingest steps, anything depending on the output of that step (the DRS ID) will fail unless we provide some dummy default. We need the secondary output so other tasks can check it to see if they should even bother using the DRS output.

Reply:
okay so I'd want to figure out how to set a default arg for the DRS ID, and do something like this https://github.com/argoproj/argo/blob/master/examples/output-parameter.yaml to have two output parameters?

Reply:
Yes, you'll want two outputs at the DAG layer, with a `default` on the DRS ID one. The RC3 that just deployed to dev should actually use the `default` now instead of failing (yay bleeding edge!). Here's an example of threading through output parameters: https://github.com/DataBiosphere/clinvar-ingest/blob/master/orchestration/templates/diff-table.yaml#L137-L144